staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1338,3 +1500,116 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
 	}
 	return models, nil
 }
+
+func (r *crdHandler) updateStorageVersionFor(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
+	gr := schema.GroupResource{
+		Group:    crd.Spec.Group,
+		Resource: crd.Spec.Names.Plural,
+	}
+	storageVersion, err := apiextensionshelpers.GetCRDStorageVersion(crd)
+	if err != nil {
+		// This should never happened if crd is valid, which is true since we
+		// only update storage version for CRDs that have been written to the
+		// storage
+		return err
+	}
+	encodingVersion := crd.Spec.Group + "/" + storageVersion
+	var decodableVersions []string
+	for _, v := range crd.Spec.Versions {
+		decodableVersions = append(decodableVersions, crd.Spec.Group+"/"+v.Name)
+	}
+
+	appendOwnerRefFunc := func(sv *apiserverinternalv1alpha1.StorageVersion) error {
+		ref := metav1.OwnerReference{
+			APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
+			Kind:       "CustomResourceDefinition",
+			Name:       crd.Name,
+			UID:        crd.UID,
+		}
+		for _, r := range sv.OwnerReferences {
+			if r == ref {
+				return nil
+			}
+		}
+		sv.OwnerReferences = append(sv.OwnerReferences, ref)
+		return nil
+	}
+	return storageversion.UpdateStorageVersionFor(
+		ctx,
+		r.storageVersionClient,
+		r.apiserverID,
+		gr,
+		encodingVersion,
+		decodableVersions,
+		appendOwnerRefFunc)
+}
+
+// getOrCreateStorageVersionUpdateQueue_locked returns the channel for the given UID,
+// or create a new one and a new goroutine if necessary. The goroutine keeps processing
+// requests until the channel is closed.
+// NOTE: caller should hold the customStorageLock until the write to the returned channel is finished,
+// to make sure the storage version update is in sync with the underlying storage change.
+func (r *crdHandler) getOrCreateStorageVersionUpdateQueue_locked(uid types.UID) chan *storageVersionUpdateRequest {
+	if queue, ok := r.storageVersionUpdateQueues[uid]; ok {
+		return queue.q
+	}
+
+	queue := make(chan *storageVersionUpdateRequest, maxStorageVersionUpdateRequests)
+	ctx, cancel := context.WithCancel(context.TODO())
+	r.storageVersionUpdateQueues[uid] = &storageVersionUpdateQueue{
```

Comment:
could storageVersionUpdateQueues be something completely independent from crdHandler ? If this go routine management was done in a self-contained unit with a clear interface, the code would become considerably simpler.